core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -17,27 +17,78 @@
-import org.apache.kafka.common.utils.Time
-
-class Replica(val brokerId: Int,
-              val topicPartition: TopicPartition,
-              time: Time = Time.SYSTEM,
-              initialHighWatermarkValue: Long = 0L,
-              @volatile var log: Option[Log] = None) extends Logging {
-  // the high watermark offset value, in non-leader replicas only its message offsets are kept
-  @volatile private[this] var highWatermarkMetadata = new LogOffsetMetadata(initialHighWatermarkValue)
+
+abstract class Replica(val brokerId: Int,
+                       val topicPartition: TopicPartition) extends Logging {
+
+  // lastCaughtUpTimeMs is the largest time t such that the offset of most recent FetchRequest from this follower >=
+  // the LEO of leader at time t. This is used to determine the lag of this follower and ISR of this partition.
+  @volatile private[this] var _lastCaughtUpTimeMs = 0L
+
+  def logStartOffset: Long
+  def logEndOffsetMetadata: LogOffsetMetadata
+
+  def logEndOffset: Long = logEndOffsetMetadata.messageOffset
+
+  def lastCaughtUpTimeMs: Long = _lastCaughtUpTimeMs
+
+  def lastCaughtUpTimeMs_=(lastCaughtUpTimeMs: Long) {
+    _lastCaughtUpTimeMs = lastCaughtUpTimeMs
+    trace(s"Setting log caught offset time for replica $brokerId, partition $topicPartition to $lastCaughtUpTimeMs")
+  }
+
+  def resetLastCaughtUpTime(curLeaderLogEndOffset: Long, curTimeMs: Long, lastCaughtUpTimeMs: Long): Unit = {
+    throw new KafkaException(
+      s"Method not implemented for partition $topicPartition and broker id $brokerId")
+  }
+
+  def updateLogReadResult(logReadResult: LogReadResult): Unit = {
+    throw new KafkaException(
+      s"Method not implemented for partition $topicPartition and broker id $brokerId")
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case other: Replica => brokerId == other.brokerId && topicPartition == other.topicPartition
+    case _ => false
+  }
+
+  override def hashCode: Int = 31 + topicPartition.hashCode + 17 * brokerId
+}
+
+class LocalReplica(brokerId: Int,
+                   topicPartition: TopicPartition,
+                   @volatile var log: Log) extends Replica(brokerId, topicPartition) {
+  def logStartOffset: Long =
+    log.logStartOffset
+
+  def logEndOffsetMetadata: LogOffsetMetadata =
+    log.logEndOffsetMetadata
+
+  override def toString: String = {
+    val replicaString = new StringBuilder
+    replicaString.append("Replica(replicaId=" + brokerId)
+    replicaString.append(s", topic=${topicPartition.topic}")
+    replicaString.append(s", partition=${topicPartition.partition}")
+    replicaString.append(s", isLocal=true")
+    replicaString.append(s", highWatermark=$log.highWatermark")
+    replicaString.append(s", lastStableOffset=$log.lastStableOffset")
+    replicaString.append(")")
+    replicaString.toString
+  }
+}
+
+class RemoteReplica(brokerId: Int,
+                    topicPartition: TopicPartition) extends Replica(brokerId, topicPartition) {
```

Comment:
nit: We should make this `final`.

Reply:
Done.

Reply:
Turned out that easymock cannot mock final classes, so ended up removing this and for LocalReplica.

Reply:
No longer needed as Local/Remote Replica is gone after latest push.